src/parse_impl.rs

```diff
@@ -0,0 +1,173 @@
+use crate::parse_fn::consume_fn;
+use crate::parse_utils::{consume_attributes, consume_stuff_until, consume_vis_marker};
+use crate::types::{Constant, ImplBody, ImplMember, TypeDefinition, ValueExpr};
+use crate::types_edition::GroupSpan;
+use crate::{Attribute, TyExpr, VisMarker};
+use proc_macro2::{Group, Ident, TokenTree};
+use std::iter::Peekable;
+
+type TokenIter = Peekable<proc_macro2::token_stream::IntoIter>;
+
+pub(crate) fn consume_for(tokens: &mut TokenIter) -> Option<Ident> {
+    if let Some(TokenTree::Ident(ident)) = tokens.peek() {
+        if ident == "for" {
+            let tk_for = ident.clone();
+            tokens.next();
+            return Some(tk_for);
+        }
+    }
+
+    None
+}
+
+pub(crate) fn parse_impl_members(token_group: Group) -> ImplBody {
+    let mut members = vec![];
+
+    let mut tokens = token_group.stream().into_iter().peekable();
+    loop {
+        if tokens.peek().is_none() {
+            break;
+        }
+
+        let attributes = consume_attributes(&mut tokens);
+        let vis_marker = consume_vis_marker(&mut tokens);
+
+        let item: ImplMember = if let Some(TokenTree::Ident(ident)) = tokens.peek() {
+            match ident.to_string().as_str() {
+                "type" => {
+                    let assoc_ty = consume_assoc_ty(&mut tokens, attributes, vis_marker);
+                    ImplMember::AssocTy(assoc_ty)
+                }
+                "const" => {
+                    // needs 2-lookahead to differentiate `const IDENT` and `const fn`
```

Comment:
I don't like this hack.

If nothing else, it doesn't handle the case where you have `const async fn foobar()`.

I think it would be better to first call `consume_fn_qualifiers`, then check for the fn keyword, and if you can't find it, "fall back" to the const case. Extract the const qualifier, throw the other qualifiers away, keep going. (Though it would give ugly code too.)